force-app/classes/src/Matcher.cls

```diff
@@ -49,17 +53,27 @@ public class Matcher {
     public Boolean matches(final Object callArgument) {
       return callArgument == this.callArgumentToMatch;
     }
+
+    override public String toString() {
+      return callArgumentToMatch + '';
+    }
   }
 
   private class JSONMatcher implements ArgumentMatcher {
-    private String callArgumentToMatch;
+    private Object callArgumentToMatch;
+    private String jsonValue;
 
     public JSONMatcher(final Object callArgumentToMatch) {
-      this.callArgumentToMatch = JSON.serialize(callArgumentToMatch);
+      this.callArgumentToMatch = callArgumentToMatch;
+      this.jsonValue = JSON.serialize(callArgumentToMatch);
     }
 
     public boolean matches(final Object callArgument) {
-      return this.callArgumentToMatch == JSON.serialize(callArgument);
+      return this.jsonValue == JSON.serialize(callArgument);
+    }
+
+    override public String toString() {
+      return 'json(' + callArgumentToMatch + ')';
```

Comment:
I started first by displaying the json and I felt like it wasn't a good idea for the sake of debugging because you'll have a json payload for params expected and raw values in the call history, so it becomes hard to spot a difference.